### PR TITLE
Rename State.config to widget everywhere

### PR DESCRIPTION
**Pull request by xster**
#7015

For your viewing pleasure.

https://travis-ci.org/flutter/flutter/builds/219488495 tested things ran ok without adding back the backward compatibility shim.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -3199,15 +3210,16 @@ class StatefulElement extends ComponentElement {
...
+    // let authors call setState from within didUpdateWidget without triggering
     // asserts.
     _dirty = true;
-    _state._config = widget;
+    _state._widget = widget;
     try {
       _debugSetAllowIgnoredCallsToMarkNeedsBuild(true);
-      _state.didUpdateConfig(oldConfig);
+      _state.didUpdateWidget(oldWidget);
+      _state.didUpdateConfig(oldWidget);
```
Comment (xster) [inline packages/flutter/lib/src/widgets/framework.dart]: Couldn't figure out how to disable the linter on this line

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -734,9 +734,14 @@ abstract class State<T extends StatefulWidget> {
...
-  /// then call [didUpdateConfig], passing the old configuration as an argument.
-  T get config => _config;
-  T _config;
+  /// then call [didUpdateWidget], passing the old configuration as an argument.
+  T get widget => _widget;
+  T _widget;
+
+  /// Deprecated getter for [widget]. Use the [widget] getter instead. This
+  /// getter will be removed on 2017/04/20.
+  @Deprecated("Will be removed 2017/04/20")
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: I recommend the following message instead:
`@Deprecated('The "config" property is now called "widget". Please migrate now, as we will remove this feature soon.')`

...or just remove it. It doesn't make any real difference if we cause the apocalypse today or on the 20th.

File: packages/flutter/lib/src/widgets/implicit_animations.dart [hunk 3]
```diff
@@ -203,7 +203,7 @@ abstract class AnimatedWidgetBaseState<T extends ImplicitlyAnimatedWidget> exten
   /// 1. Call the visitor callback with three arguments, the first argument
   /// being the current value of the Tween<T> object that represents the
   /// tween (initially null), the second argument, of type T, being the value
-  /// on the Widget (config) that represents the current target value of the
+  /// on the Widget (widget) that represents the current target value of the
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/implicit_animations.dart]: This parenthetical doesn't make much sense any more.

File: packages/flutter/test/widgets/focus_test.dart [hunk 4]
```diff
@@ -47,8 +47,8 @@ class TestFocusableState extends State<TestFocusable> {
       child: new AnimatedBuilder(
         animation: focusNode,
         builder: (BuildContext context, Widget child) {
-          // print('focusNode.hasFocus = ${focusNode.hasFocus} ${focusNode.hashCode} ${focusNode.hasFocus ? config.yes : config.no}');
-          return new Text(focusNode.hasFocus ? config.yes : config.no);
+          // print('focusNode.hasFocus = ${focusNode.hasFocus} ${focusNode.hashCode} ${focusNode.hasFocus ? widget.yes : widget.no}');
```
Comment (Hixie) [inline packages/flutter/test/widgets/focus_test.dart]: remove this commented out code

File: packages/flutter/test/material/dialog_test.dart [hunk 5]
```diff
@@ -90,11 +90,11 @@ void main() {
     await tester.pump(const Duration(seconds: 1));
 
     final StatefulElement widget = tester.element(find.byType(Material).last);
-    final Material materialconfig = widget.state.config;
+    final Material materialwidget = widget.state.widget;
```
Comment (Hixie) [inline packages/flutter/test/material/dialog_test.dart]: can you rename this materialWidget (capital W)?

File: packages/flutter/lib/src/material/paginated_data_table.dart [hunk 6]
```diff
@@ -302,9 +302,9 @@ class PaginatedDataTableState extends State<PaginatedDataTable> {
     } else {
       headerWidgets.add(new Expanded(child: new Text('$_selectedRowCount items selected')));
     }
-    if (config.actions != null) {
+    if (widget.actions != null) {
       headerWidgets.addAll(
-        config.actions.map<Widget>((Widget widget) {
+        widget.actions.map<Widget>((Widget widget) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/paginated_data_table.dart]: This code now has two `widget` variables in scope, which is confusing. We should rename the inner one (maybe to "action").

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: remove the call? I think you just ended up with two calls where you had one, or something.

**Comment by Hixie** [comment]
There are many, many places where we have a "widget" local variable today, where with this patch there's now shadowing going on, where the local variable is shadowing the property previously known as "config". I think we should make sure all such cases are fixed before checking this in (that was the problem that had stopped Adam from doing it when he tried).

File: packages/flutter/lib/src/widgets/framework.dart [hunk 7]
```diff
@@ -801,31 +806,37 @@ abstract class State<T extends StatefulWidget> {
...
-  /// super.didUpdateConfig(oldConfig).
+  /// super.didUpdateWidget(oldWidget).
   // TODO(abarth): Add @mustCallSuper.
   @protected
+  void didUpdateWidget(covariant T oldWidget) { }
+
+  /// Deprecated callback when the widget configuration changes.
+  ///
+  /// Use [didUpdateWidget] instead. This method will be removed on 2017/04/20.
+  @Deprecated("Will be removed 2017/04/20")
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: ditto here regarding the change suggested for @Deprecated -- either this should be a better message or we should just remove it entirely.

**Comment by xster** [comment]
Ya you're right. I thought Adam just had an IDE refactor issue. Wish there would be an analyzer check for it :S

There's 37 of them. I'll rename them.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/framework.dart]: We still want the second one for the transitional period so people who listened to the old one don't silently stop getting calls.

Comment (abarth) [inline packages/flutter/lib/src/widgets/framework.dart]: I'd just remove it.  There are already three mitigation: (1) the email, (2) the change to `config`, which will get their attention, (3) the `@override` annotation.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/framework.dart]: Ok

File: packages/flutter/lib/src/widgets/implicit_animations.dart (hunk 3 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/implicit_animations.dart]: Done

File: packages/flutter/test/widgets/focus_test.dart (hunk 4 above)
Comment (xster) [inline packages/flutter/test/widgets/focus_test.dart]: Done

File: packages/flutter/test/material/dialog_test.dart (hunk 5 above)
Comment (xster) [inline packages/flutter/test/material/dialog_test.dart]: Done

File: packages/flutter/lib/src/widgets/framework.dart (hunk 7 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/framework.dart]: Done

**Comment by xster** [comment]
Thanks for reviewing. 

Turns out on closer inspection, for subclasses of State that also has any field named 'widget', there are only 3 left. Renamed.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (xster) [inline packages/flutter/lib/src/widgets/framework.dart]: Should I just remove the @Deprecated config then? If we're going to non backward compatible change, we might as well do it once.

**Comment by Hixie** [comment]
I wouldn't bother with the @deprecation, yeah.

**Comment by Hixie** [comment]
Looking into this it seems there are still many cases of identifiers called "widget" that shadow the new getter, or that are in parts of the codebase where it's now confusing that they have the same name.

For example, the following files are examples of files with build functions with a local lambda whose parameter is named `widget` which shadows the State.widget getter:
* `examples/flutter_gallery/lib/demo/material/chip_demo.dart`
* `packages/flutter/lib/src/material/paginated_data_table.dart`
* there's lots of others

Also the `DataCell` class has a `widget` property that should probably be renamed `child` for consistency now.

**Comment by xster** [comment]
Thanks for going through them again. 

Hoping you caught the last 2 cases :)

I don't think there were any variables called 'widget' in paginated_data_table.

Went through the declarations of a 'widget' variable in the codebase again. As far as I can tell, besides those 2, all remaining variables called 'widget' are in tests, the testing framework (as static methods), stateless widgets and lower level renderers.

**Comment by Hixie** [comment]
LGTM

**Pull request merged**